ros/angel_system_nodes/angel_system_nodes/feedback_generator.py

```diff
@@ -0,0 +1,151 @@
+import time
+
+import uuid
+
+import rclpy
+from rclpy.node import Node
+from vision_msgs.msg import BoundingBox3D
+from geometry_msgs.msg import Pose, Point
+
+from angel_msgs.msg import (
+    ActivityDetection,
+    ObjectDetection3dSet,
+    TaskUpdate,
+    AruiUpdate,
+    AruiObject3d
+)
+
+
+class FeedbackGenerator(Node):
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+
+        self.declare_parameter("activity_detector_topic", "ActivityDetections")
+        self.declare_parameter("object_detection_topic", "ObjectDetections3d")
+        # TODO: task needs to be updated to emit TaskNode
+        self.declare_parameter("task_monitor_topic", "TaskUpdates")
+        self.declare_parameter("arui_update_topic", "AruiUpdates")
+
+        self._activity_detector_topic = self.get_parameter("activity_detector_topic").get_parameter_value().string_value
+        self._object_detection_topic = self.get_parameter("object_detection_topic").get_parameter_value().string_value
+        self._task_monitor_topic = self.get_parameter("task_monitor_topic").get_parameter_value().string_value
+        self._arui_update_topic = self.get_parameter("arui_update_topic").get_parameter_value().string_value
+
+        # logger
+        self.log = self.get_logger()
+        self.log.info(f"Activity detector topic: {self._activity_detector_topic}")
+        self.log.info(f"Object detection topic: {self._object_detection_topic}")
+        self.log.info(f"Task monitor topic: {self._task_monitor_topic}")
+        self.log.info(f"AruiUpdate topic: {self._arui_update_topic}")
+
+        # subscribers
+        self.activity_subscriber = self.create_subscription(
+            ActivityDetection,
+            self._activity_detector_topic,
+            self.activity_callback,
+            1
+        )
+
+        self.object_detector_subscriber = self.create_subscription(
+            ObjectDetection3dSet,
+            self._object_detection_topic,
+            self.object_callback,
+            1
+        )
+
+        self.task_monitor_subscriber = self.create_subscription(
+            TaskUpdate,
+            self._task_monitor_topic,
+            self.task_callback,
+            1
+        )
+
+        # publisher
+        self.arui_update_publisher = self.create_publisher(
+            AruiUpdate,
+            self._arui_update_topic,
+            1
+        )
+
+        # message
+        self.arui_update_message = AruiUpdate()
+
+        # detection uuids
+        self.uuids = dict()
+
+    def publish_update(self):
+        self.arui_update_message.header.stamp = self.get_clock().now().to_msg()
+        self.log.info(f"Publishing AruiUpdate: {self.arui_update_message}\n")
+        self.arui_update_publisher.publish(self.arui_update_message)
+
+    def activity_callback(self, activity):
+        self.arui_update_message.latest_activity = activity
+        self.arui_update_message.header.frame_id = activity.header.frame_id
```

Comment:
Should the `AruiUpdate` message header frame ID be changed in every callback? I thought it should always be the same, like "ARUI update".

Reply:
Maybe @Purg can answer.

Reply:
I missed that. In the current use where there is one global message that is constantly being updated, not we probably don't want to update the frame ID all the time as the ARUI Update message "frame of reference" doesn't change over time (it's probably an arbitrary value, like "system", that we won't pay attention to) so it could be set once. However, this sort of ties into my race-condition comment: if you put a lock around every message modification and sending, then the above still holds.